### PR TITLE
General modals refactor

### PR DESCRIPTION
**Pull request by samelliott89**
### PR Checklist (check all)

* [x] Updated `CHANGELOG.md` to describe the included fixes and changes made
* [x] Included issue number on the description of the PR
* [x] Commented on the relevant issue thread about this PR

### Issue number 

#238 

### Added
- New util function for transaction prep
- General componetisation of the modals
- General improvement to readability/maintainability of the modals

### Removed
- Duplicative code

### CODE REVIEW

**Comment by samelliott89** [comment]
@pedrouid I wanted to push this up as WIP to get some feedback on some of the early thinking here. Just basic stuff, but like we spoke about, starting to componentize some of the modals (I used the Donate modal first just to show) with the intention to roll out to other modals. This is first pass (i know some functions are poorly named etc). If you're happy with this scaffold I can tidy this up and either 1. Put in a PR for just Donate Modal or 2. Roll out to Send / Receive etc

**Comment by samelliott89** [comment]
@pedrouid any feedback on this so far?

**Comment by ricburton** [comment]
@pedrouid is going to focus on WalletConnect. @jinchung will take a look at this :)

Richard

**Comment by ricburton** [comment]
🙌 Preview this pull request.

Built with commit 2f4874caac932677072bf67e44efe3cd00422866

https://deploy-preview-257--balance-manager.netlify.com

**Comment by samelliott89** [comment]
I didn't want this PR to become too big so I've put a hard stop on it here.

**Comment by ricburton** [comment]
Great. @jinchung is in shapeshift mode then will give this the attention it deserves.

**Comment by jinchung** [comment]
thanks a lot @samelliott89 ! 

Resolved the merge conflicts and pulled in your changes to latest.
Had a separate question, but I will comment elsewhere!

File: src/modals/index.js [hunk 1]
```diff
@@ -45,15 +49,26 @@ const StyledContainer = styled.div`
   justify-content: center;
 `;
 
+const reduxProps = ({ modal }) => ({
+  modal: modal.modal,
+});
+
 class Modal extends Component {
+  propTypes = {
```
Comment (jinchung) [inline src/modals/index.js:57]: @samelliott89 I noticed I was seeing a warning: propTypes was defined as an instance property on Modal. Use a static property to define propTypes instead.

Comment (samelliott89) [reply src/modals/index.js:57]: I think we can add `static propTypes = {}` and that should fix, just tested locally.

**Comment by samelliott89** [comment]
hi @jinchung ! I would have liked to have discussed the PR a bit more before merging, but if you are happy with the code proposal than all good!

There are a few things I wanted to run past you as I have made significant changes to some critical code (Sending and Receiving Modals). I'm not sure what your QA process is before deploying to production, but just want to make sure we test a few of those flows.

Some of more significant changes were some of these new util functions:

```
export const transactionData = (accountInfo, assetAmount, gasPrice) => {
  const ethereum = getEth(accountInfo.assets);
  const balanceAmount = ethereum.balance.amount;
  const balance = convertAmountFromBigNumber(balanceAmount);
  const requestedAmount = convertNumberToString(assetAmount);
  const txFeeAmount = gasPrice.txFee.value.amount;
  const txFee = convertAmountFromBigNumber(txFeeAmount);
  const amountWithFees = add(requestedAmount, txFee);

  return {
    ethereum,
    balanceAmount,
    balance,
    requestedAmount,
    txFeeAmount,
    txFee,
    amountWithFees,
  };
};
```
And this fee calculator
```
export const calcTxFee = (gasPrices, gasPriceOption) => {
  let nativeFee = '$0.00';
  let txFee = '0.000 ETH';

  const option = gasPrices[gasPriceOption];
  const isAvailable = option && option.txFee;

  if (isAvailable) {
    nativeFee = gasPrices[gasPriceOption].txFee.native.value.display;
    txFee = gasPrices[gasPriceOption].txFee.value.display;
  }

  return `${nativeFee} - (${txFee})`;
};

```
Most of the code is essentially just consolidation of duplicative code, in an attempt to componentize things a bit more so that when we do start writing tests. I did a limited amount of testing as I do not have a tonne of assets on my metamask wallet.

File: src/modals/index.js (hunk 1 above)
Comment (jinchung) [reply src/modals/index.js:57]: Hey @samelliott89  - yes you're right, it just needs to be declared as static. I'll add that. Thanks a lot! Cheers, will add these

**Comment by jinchung** [comment]
Got it - thanks @samelliott89 will take a look at the util functions. Sorry for the delayed response.

**Comment by jinchung** [comment]
Hey @samelliott89  - thanks for bringing those up. I did a couple small tweaks shortly afterwards but haven't updated you yet. For the first one, there were some differences in the token logic that used transactionData (I can't remember off the top of my head now, but something about not using the amountWithFees given that for tokens you don't want to mix up the amount in token plus fees in ETH, etc). For the second one, I noticed there was a slight difference in the logic so I just adjusted accordingly but kept these as separate utils since it reduces code repetition overall. Cheers! Thanks for your contribution!

**Pull request merged**